# Data Science project basics

> This is my first data science project ever developed. With this, an attempt was made to learn about the data science.
> This project contains various data science methods like data extraction, data cleaning, graphs and many more.

## Data exploration;

Below code is used to explore the data from the data sets.
* There are two datasets. 
    * Google play store dataset.
    * Apple store dataset.

In [1]:
from csv import reader

### opening playstore data set
playStore_file = open('googleplaystore.csv')
read_playStore_file = reader(playStore_file)
playStore = list(read_playStore_file)

playStore_header = playStore[0]
playStore = playStore[1:]

### opening applestore data set
appleStore_file = open('applestore.csv')
read_appleStore_file = reader(appleStore_file)
appleStore = list(read_appleStore_file)

appleStore_header = appleStore[0]
appleStore = appleStore[1:]


### Exploring the data
Here we write a function, __explore_data__ which takes following arguments:
* __dataset__ : This contains the dataset as input that is extracted from the above step.
* __start__ : Start point of the slice.
* __end__ : End point of the slice.
* __rows_and_colums = False__: This is by default false. If set true then number of rows and columns of the dataset are displayed.

In [2]:
def explore_data(dataset, start, end, rows_and_columns = False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n')
    if rows_and_columns:
        print("Number of rows in the dataset: ", len(dataset))
        print("Number of columns in the dataset: ", len(dataset[0]))

In [3]:
explore_data(playStore, 1, 5, rows_and_columns = False)

['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']




In [4]:
explore_data(appleStore, 1, 3, rows_and_columns = True)

['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows in the dataset:  7197
Number of columns in the dataset:  16


### Data Cleaning
Now we shall clean the data. __Data Cleaning__ is the step where data is cleaned inorder to get the accurate results. In this cleaning data phase, following steps are done:
* Inaccurate data is detected. We either correct it or remove it.
* Duplicate data is detected. These are removed.

In the below step, we delete the row with missing data. In our case, record 10472 has a wrong entry. So we delete that row. __(Make sure the we execute the del statement only once.)__ This deletion step is in __ln[6]__

In [5]:
print(playStore_header)
print('\n')
print(playStore[1])
print('\n')
print(playStore[10472])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [6]:
#del(playStore[10472])
print(playStore[10472])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In the above piece of code, we removed the wrong data entry. Now let us try removing the duplicate values.
Here in this example, we are trying to remove the duplicate value __name = Instagram__ from the googleplaystore dataset.

In [7]:
for app in playStore:
    name = app[0]
    if name == "Instagram":
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


### Finding duplicate entries in the data
So if observed, there are 4 duplicate __Instagram__ apps. And also it is said in the problem statement [here](https://github.com/dataquestio/solutions/blob/master/Mission350Solutions.ipynb) (see below line 6), there are in total 1181 duplicate apps in the __playStore dataset.__ 
in the below step, let us separate duplicate apps from the unique apps.

In [8]:
unique_apps = []
duplicate_apps = []

for app in playStore:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(app[0])
    else:
        unique_apps.append(app[0])
        
print("Total number of duplicate apps: ", len(duplicate_apps))
print("Total number of unique apps: ",len(unique_apps))

Total number of duplicate apps:  1181
Total number of unique apps:  9660


As mentioned, there are 1181 duplicate apps in the playstore. Hence the next step would be the removal of duplicate entries.

### Removing duplicate entries
In order to get the accurate analysis of the data, we shall have to remove all the duplicate entries from the dataset.
But rather than removing the duplicate entries randomly, there is a better way to remove them.

If we observe the results of the duplicate app __Instagram__, the main difference occurs in the 4th position of the entry, that corresponds to the rating os the application. This states that the ratings of the app were collected at different times. Hence the better way of removing this entry is to keep that entry whose rating is highest and discard the remaining values. This way, we get more accurate results. Let's do this as the below steps:
* Create a dictionary where each key is a unique app name, and the value is the highest number of reviews of that app
* Use the dictionary to create a new data set, which will have only one entry per app (and we only select the apps with the highest number of reviews)

Let's start creating the dictionary

In [9]:
reviews_max = {}


for app in playStore:
    name = app[0]
    n_reviews = app[3]
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
        
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

Hence in the above code, we did create a dictionary that holds the data excluding the duplicate entries from the dataset. 
So it is clear that the number of duplicate entries in the data set is 1181. So the difference obtained after removing the duplicate apps from all the apps present in the playStore data set should be equal to the length of the dictionary. Let's find out:

In [10]:
print("Expected length: ", len(playStore) - 1181)
print("Actual length: ", len(reviews_max))

Expected length:  9660
Actual length:  9660


Now, let us use the dictionary to remove the duplicate values from the dataset __playStore__. We will obtained a __clean dataset__ after performing this activity.

In [11]:
clean_playStore = []
already_added = []

for app in playStore:
    name = app[0]
    reviews = app[3]
    
    if (reviews_max[name] == reviews) and (name not in already_added):
        clean_playStore.append(app)
        already_added.append(name)

In [12]:
explore_data(clean_playStore, 0, 4, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


Number of rows in the dataset:  9660
Number of columns in the dataset:  13


In the above steps, we have successfully removed the duplicate apps from the dataset using dictionary and created a clean data set called __clean_playStore__.

The main motto of the project is to analyse the __English app sales in both playStore and appleStore.__ So now it is time to remove non english apps from the clean data set.

### Removing non-english apps from the dataset.
There are many non-english apps in the clean dataset we obtained from the above step. This can be done using python's inbuilt function called __ord()__. This function returns the unicode of a symbol in python. All the unicodes above 127 are special characters that state non-english letters. Further more information can be found [here](https://en.wikipedia.org/wiki/List_of_Unicode_characters)

It is given that entries 4415 and 7940 in clean_playStore datasets are non-English apps as below:

In [13]:
print(clean_playStore[4415][0])
print(clean_playStore[7941][0])

中国語 AQリスニング
لعبة تقدر تربح DZ


#### Part 1 (Removing non-english apps using ord())
We are not interested in the apps like above. So let us create a function to identify non-english strings as below:

In [14]:
def is_english(string):
    
    for character in string:
        if(ord(character) > 127):
            return False
        
    return True

Let us test the above function with non-english apps

In [15]:
print(is_english("Instagram"))
print(is_english("中国語 AQリスニング"))

True
False


Now that we see, the above function works fine. But in many cases english apps do have a special character. Whose ascii values and unicodes might defer with the above snippet's unicodes. Look below for such exceptional cases.

In [16]:
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

print(ord('™'))
print(ord('😜'))

False
False
8482
128540


So if this is the case, we will loose the data of some important applications from the dataset. So, we shall now takecare of such special cases.
#### Part 1 (Minimizing data loss)
In this step, we shall return false if there are more than 3 invalid ASCII characters.

In [17]:
def is_english_2(string):
    non_ascii = 0
    for character in string:
        if(ord(character) > 127):
            non_ascii += 1
        
    if non_ascii > 3:
        return False
    else:
        return True

In [18]:
print(is_english_2('Docs To Go™ Free Office Suite'))
print(is_english_2('Instachat 😜'))

True
True


The function is still not perfect, and very few non-English apps might get past our filter, but this seems good enough at this point in our analysis — let's not spend too much time on this optimization now.

Below, we use the __is_english_2( )__ function to filter out the non-English apps for both data sets:

In [19]:
playStore_english = []
appleStore_english = []

for app in clean_playStore:
    if is_english_2(app[0]):
        playStore_english.append(app)
        
for app in appleStore:
    if is_english_2(app[1]):
        appleStore_english.append(app)

explore_data(playStore_english, 0, 3, True)
explore_data(appleStore_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows in the dataset:  9615
Number of columns in the dataset:  13
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '

Hence, now in total we are left with 9615 english playStore apps, and 6183 english appleStore apps.

### Isolate free apps
we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. Our data sets contain both free and non-free apps, and we'll need to isolate only the free apps for our analysis. Below, we isolate the free apps for both our data sets.

In [20]:
playStore_final = []
appleStore_final = []

for app in playStore_english:
    price = app[7]
    if price == '0':
        playStore_final.append(app)

for app in appleStore_english:
    price = app[4]
    if price == '0.0':
        appleStore_final.append(app)
        
print("length of the final playStore app: ",len(playStore_final))
print("length of the final appleStore app: ",len(appleStore_final))

length of the final playStore app:  8862
length of the final appleStore app:  3222


### Most Common Apps by genre

We are interested in the apps that are most common in both playStore and appleStore. So let us create a frequency table function and then display function.
* The frequency table shall find the percentages of the most common apps.
* Another function to display the percentages in deschending order.


In [21]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    table_percentages = {}
    for key in table:
        percentage = (table[key]/total)*100
        table_percentages[key] = percentage

    return table_percentages

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

#### Part Three
Now let us examine prime_genre column(index = -5 i,e. 5th column from last) in appStore dataset.

In [22]:
display_table(appleStore_final , -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


### Conclusions
* We can see that in English free apps in ios store, more than 58% apps are games and so on.
* All other apps that come under productivity constitute less than 10%

This inferences might be wrong. It doesn't mean that if the #apps are more in the app store, it means the user base is high in those apps. Games/Entertainment apps in number might be more but the user base might be more for other apps. So we can not simply make conclusions based on the number of apps.

Let's continue by examining the Genres and Category columns of the Google Play data set (two columns which seem to be related).

In [23]:
display_table(playStore_final, 1) # Category

FAMILY : 18.934777702550214
GAME : 9.693071541412774
TOOLS : 8.451816745655607
BUSINESS : 4.5926427443015125
LIFESTYLE : 3.9043105393816293
PRODUCTIVITY : 3.8930264048747465
FINANCE : 3.7011961182577298
MEDICAL : 3.5206499661475967
SPORTS : 3.39652448657188
PERSONALIZATION : 3.3175355450236967
COMMUNICATION : 3.238546603475513
HEALTH_AND_FITNESS : 3.080568720379147
PHOTOGRAPHY : 2.945159106296547
NEWS_AND_MAGAZINES : 2.798465357707064
SOCIAL : 2.663055743624464
TRAVEL_AND_LOCAL : 2.335815842924848
SHOPPING : 2.2455427668697814
BOOKS_AND_REFERENCE : 2.143985556307831
DATING : 1.8618821936357481
VIDEO_PLAYERS : 1.7941773865944481
MAPS_AND_NAVIGATION : 1.399232678853532
FOOD_AND_DRINK : 1.2412547957571656
EDUCATION : 1.1735499887158656
ENTERTAINMENT : 0.9591514330850823
LIBRARIES_AND_DEMO : 0.9365831640713158
AUTO_AND_VEHICLES : 0.9252990295644324
HOUSE_AND_HOME : 0.8237418190024826
WEATHER : 0.8011735499887158
EVENTS : 0.7109004739336493
PARENTING : 0.6544798013992327
ART_AND_DESIGN : 0.